data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetBlockHeaders.java

```diff
@@ -79,29 +104,29 @@ public GetBlockHeaders(
       })
   @Override
   public void handle(@NotNull final Context ctx) throws Exception {
-    final Map<String, List<String>> queryParameters = ctx.queryParamMap();
-
-    final Optional<Bytes32> parentRoot =
-        getParameterValueAsBytes32IfPresent(queryParameters, PARENT_ROOT);
-    final Optional<UInt64> slot = getParameterValueAsUInt64IfPresent(queryParameters, SLOT);
-    try {
-      ctx.future(
-          chainDataProvider
-              .getBlockHeaders(parentRoot, slot)
-              .thenApplyChecked(jsonProvider::objectToJSON)
-              .exceptionallyCompose(error -> handleError(ctx, error)));
-    } catch (final IllegalArgumentException e) {
-      ctx.status(SC_BAD_REQUEST);
-      ctx.json(jsonProvider.objectToJSON(new BadRequest(e.getMessage())));
-    }
+    adapt(ctx);
   }
 
-  private SafeFuture<String> handleError(final Context ctx, final Throwable error) {
-    final Throwable rootCause = Throwables.getRootCause(error);
-    if (rootCause instanceof IllegalArgumentException) {
-      ctx.status(SC_BAD_REQUEST);
-      return SafeFuture.of(() -> BadRequest.badRequest(jsonProvider, error.getMessage()));
-    }
-    return SafeFuture.failedFuture(error);
+  @Override
+  public void handleRequest(RestApiRequest request) throws JsonProcessingException {
+    final Optional<Bytes32> parentRoot = request.getOptionalQueryParameter(PARENT_ROOT_PARAMETER);
+    final Optional<UInt64> slot = request.getOptionalQueryParameter(SLOT_PARAMETER);
+
+    final SafeFuture<BlockHeadersResponse> future =
+        chainDataProvider.getBlockHeaders(parentRoot, slot);
+
+    request.respondAsync(
+        future
+            .thenApplyChecked(AsyncApiResponse::respondOk)
+            .exceptionallyCompose(
+                error -> {
+                  final Throwable rootCause = Throwables.getRootCause(error);
+                  if (rootCause instanceof IllegalArgumentException) {
+                    return SafeFuture.of(
+                        () ->
+                            AsyncApiResponse.respondWithError(SC_BAD_REQUEST, error.getMessage()));
```

Comment:
do we have a test that gets into here? I'm not sure whether this will work...

Reply:
I successfully got back a bad request response when running it in Swagger UI. But could add a test.

Reply:
I guess I just wasn't sure if the 'of' worked in that way, but if it's working thats good :)

Reply:
The `.of` will immediately evaluate it so unless there's an exception that may be thrown it would be simpler to just say `SafeFuture.completedFuture(AsyncApiResponse.respondWithError(....`